### PR TITLE
cgo,native: implement dlapmt and dlapll, and tests

### PR DESCRIPTION
**Pull request by kortschak**
@btracey @vladimir-ch Please take a look.

Updates #218.

### CODE REVIEW

File: native/dlapmt.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	checkMatrix(m, n, x, ldx)
+	if len(k) != n {
+		panic(badKperm)
+	}
+
+	if n <= 1 {
+		return
+	}
+
+	for i, v := range k {
```
Comment (kortschak) [inline native/dlapmt.go]: I don't like all the 0 to 1 indexing fiddling here, but I don't see a better way to do this.

File: native/dlapmt.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			}
+			j := i + 1
+			in := -v
+			k[j-1] = in
+			for {
+				if k[in-1] > 0 {
+					break
+				}
+
+				// for ii := 0; ii < m; ii++ {
```
Comment (kortschak) [inline native/dlapmt.go]: These commented-out section are just to show what I have replaced with the `Dswap` call. I can leave them in or take them out.

File: native/general.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ const (
 	badIhi          = "lapack: ihi out of range"
 	badIpiv         = "lapack: insufficient permutation length"
 	badJob          = "lapack: bad Job"
+	badKperm        = "lapack: incorrect permutation length"
```
Comment (kortschak) [inline native/general.go]: Same here.

File: native/dlapmt.go (hunk 1 above)
Comment (btracey) [inline native/dlapmt.go]: Before I review, this code is a true permutation, right? It's not just specifying a set of row swaps like other Lapack code.

Comment (btracey) [inline native/dlapmt.go]: I'm asking because I'm pretty confused by all the negations.

**Comment by btracey** [comment]
Dlapll implementation looks good.

File: cgo/lapack.go [hunk 4]
```diff
@@ -168,6 +169,46 @@ func (impl Implementation) Dlacpy(uplo blas.Uplo, m, n int, a []float64, lda int
 	lapacke.Dlacpy(uplo, m, n, a, lda, b, ldb)
 }
 
+// Dlapmt rearranges the columns of the m×n matrix x as specified by the
```
Comment (vladimir-ch) [inline cgo/lapack.go]: I (and maybe not only I) usually use a capital letter to denote the slice+ld matrix pair. It's up to you. Also elsewhere in this PR.

File: cgo/lapack.go [hunk 5]
```diff
@@ -168,6 +169,46 @@ func (impl Implementation) Dlacpy(uplo blas.Uplo, m, n int, a []float64, lda int
...
+// permutation k[0], k[1], ..., k[n-1] of the integers 0, ..., n-1.
+//
+// If forward is true a forward permutation is performed:
+//
+//  x[*, k[j]] is moved x[*, j] for j = 0, 1, ..., n-1.
+//
+// otherwise a  backward permutation is performed:
```
Comment (vladimir-ch) [inline cgo/lapack.go]: Two spaces.

File: cgo/lapack.go [hunk 6]
```diff
@@ -168,6 +169,46 @@ func (impl Implementation) Dlacpy(uplo blas.Uplo, m, n int, a []float64, lda int
...
+//
+//  x[*, j] is moved to x[*, k[j]] for j = 0, 1, ..., n-1.
+//
+// k must have length n, otherwise Dlapmt will panic otherwise. k is zero-indexed.
```
Comment (kortschak) [inline cgo/lapack.go]: Done.

File: cgo/lapack.go (hunk 4 above)
Comment (kortschak) [inline cgo/lapack.go]: That is something I have also previously done. Fixed.

File: cgo/lapack.go (hunk 5 above)
Comment (kortschak) [inline cgo/lapack.go]: Fixed.

File: native/dlapmt.go (hunk 1 above)
Comment (kortschak) [inline native/dlapmt.go]: This is a true permutation. I remember writing a more concise permutation routine a couple of years back, but I would have to dig it up and I don't know how will the reverse would be implemented.

The negatives indicate whether the `j` index has been done (negatives are not yet done).

File: native/dlapmt.go [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+	if forward {
+		for j, v := range k {
+			if v >= 0 {
+				continue
+			}
+			k[j] = -v
+			i := -v - 1
+			for k[i] < 0 {
+				// for ii := 0; ii < m; ii++ {
```
Comment (kortschak) [inline native/dlapmt.go]: Do you want these commented section kept or removed?

File: native/dlapmt.go (hunk 1 above)
Comment (kortschak) [inline native/dlapmt.go]: I've looked through the approach I used and it requires a []bool to be allocated. This does not.

File: native/dlapmt.go [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+//
+// If forward is true a forward permutation is performed:
+//
+//  x[*, k[j]] is moved x[*, j] for j = 0, 1, ..., n-1.
+//
+// otherwise a  backward permutation is performed:
+//
+//  x[*, j] is moved to x[*, k[j]] for j = 0, 1, ..., n-1.
+//
+// k must have length n, otherwise Dlapmt will panic otherwise. k is zero-indexed.
```
Comment (kortschak) [inline native/dlapmt.go]: Done.

File: native/dlapmt.go [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+package native
+
+import "github.com/gonum/blas/blas64"
+
+// Dlapmt rearranges the columns of the m×n matrix X as specified by the
+// permutation k[0], k[1], ..., k[n-1] of the integers 0, ..., n-1.
+//
+// If forward is true a forward permutation is performed:
+//
+//  X[*, k[j]] is moved X[*, j] for j = 0, 1, ..., n-1.
```
Comment (vladimir-ch) [inline native/dlapmt.go]: `is moved to`

File: cgo/lapack.go [hunk 10]
```diff
@@ -168,6 +169,46 @@ func (impl Implementation) Dlacpy(uplo blas.Uplo, m, n int, a []float64, lda int
 	lapacke.Dlacpy(uplo, m, n, a, lda, b, ldb)
 }
 
+// Dlapmt rearranges the columns of the m×n matrix X as specified by the
+// permutation k[0], k[1], ..., k[n-1] of the integers 0, ..., n-1.
+//
+// If forward is true a forward permutation is performed:
+//
+//  X[*, k[j]] is moved X[*, j] for j = 0, 1, ..., n-1.
```
Comment (vladimir-ch) [inline cgo/lapack.go]: `is moved to`

Also, I have been trying to consistently use `:` instead of `*` in my PRs.

File: native/dlapmt.go (hunk 7 above)
Comment (vladimir-ch) [inline native/dlapmt.go]: I think it is ok to remove them.

File: testlapack/dlapmt.go [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+func DlapmtTest(t *testing.T, impl Dlapmter) {
+	for ti, test := range []struct {
+		forward bool
+		k       []int
+
+		want blas64.General
+	}{
+		{
+			forward: true, k: []int{0, 1, 2},
```
Comment (vladimir-ch) [inline testlapack/dlapmt.go]: On separate lines?

Comment (kortschak) [inline testlapack/dlapmt.go]: They make a set of input parameters, so I'm OK with this.

File: native/dlapll.go [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+	// Compute the QR factorization of the N-by-2 matrix [ X Y ].
+	_, tau := impl.Dlarfg(n, x[0], x[incX:], incX)
+	a00 := x[0]
+	x[0] = 1
+
+	bi := blas64.Implementation()
+	c := -tau * bi.Ddot(n, x, incX, y, incY)
+	bi.Daxpy(n, c, x, incX, y, incY)
+	impl.Dlarfg(n-1, y[incY], y[2*incY:], incY)
```
Comment (kortschak) [inline native/dlapll.go]: Done.

File: native/dlapll.go [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+func (impl Implementation) Dlapll(n int, x []float64, incX int, y []float64, incY int) float64 {
+	checkVector(n, x, incX)
+	checkVector(n, y, incY)
+
+	if n <= 1 {
+		return 0
+	}
+
+	// Compute the QR factorization of the N-by-2 matrix [ X Y ].
+	_, tau := impl.Dlarfg(n, x[0], x[incX:], incX)
```
Comment (kortschak) [inline native/dlapll.go]: And incX != 0, so I don't need to reset x[0] to 1, since x[0:] cannot be a parameter. True?

Comment (kortschak) [inline native/dlapll.go]: Done.

File: native/dlapmt.go (hunk 9 above)
Comment (kortschak) [inline native/dlapmt.go]: Done.

File: cgo/lapack.go (hunk 10 above)
Comment (kortschak) [inline cgo/lapack.go]: Done.

File: native/dlapll.go [hunk 14]
```diff
@@ -0,0 +1,37 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import "github.com/gonum/blas/blas64"
+
+// Dlapll returns the smallest singular value of the n×2 matrix a = [ x y ].
+// The function first computes the QR factorization of a = q*r, and then computes
```
Comment (kortschak) [inline native/dlapll.go]: Done.

File: native/dlapll.go [hunk 15]
```diff
@@ -0,0 +1,37 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import "github.com/gonum/blas/blas64"
+
+// Dlapll returns the smallest singular value of the n×2 matrix a = [ x y ].
```
Comment (kortschak) [inline native/dlapll.go]: Done.

File: native/dlapll.go [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+func (impl Implementation) Dlapll(n int, x []float64, incX int, y []float64, incY int) float64 {
+	checkVector(n, x, incX)
+	checkVector(n, y, incY)
+
+	if n <= 1 {
+		return 0
+	}
+
+	// Compute the QR factorization of the N-by-2 matrix [ X Y ].
+	a00, tau := impl.Dlarfg(n, x[0], x[incX:], incX)
```
Comment (vladimir-ch) [inline native/dlapll.go]: I think you need to keep `x[0] = 1` for the `Ddot` and `Daxpy` calls below.

Comment (kortschak) [inline native/dlapll.go]: Yes. I was just protecting from change, not setting to 1. Silly.

Fixed.

Comment (kortschak) [inline native/dlapll.go]: Uncovered a foolish error in the test.

File: cgo/lapack.go (hunk 10 above)
Comment (vladimir-ch) [inline cgo/lapack.go]: Please make the change also here in cgo so that the docs are identical.

**Comment by kortschak** [comment]
@btracey are you satisfied?

**Comment by btracey** [comment]
LGTM

**Pull request merged**